docs/controllers/optimization.md

```diff
@@ -0,0 +1,224 @@
+
+
+### Reducing Number of Watched Objects
+The default `watcher::Config` will watch every object in the [Api] scope you configured:
+
+- `Api::namespaced` or `Api::default_namespaced` -> all objects in that namespace
+- `Api::all` -> all cluster scoped objects (or all objects in all namespaces)
+
+This can be limited to just a subset of namespaces, or a subset of objects using field selectors to limit to a selection of known names:
+
+```rust
+let cfg = watcher::Config::default().fields(&format!("metadata.name={name}"));
+```
+
+This can also be comma-delimited for more names, and similarly you you can also comma deliminate an exclusion list on names or namespaces:
+
+```rust
+let ignoring_system_namespaces = [
+    "carousel",
+    "cattle-fleet-system",
+    "cattle-impersonation-system",
+    "cattle-monitoring-system",
+    "cattle-system",
+    "fleet-system",
+    "gatekeeper-system",
+    "kube-node-lease",
+    "kube-public",
+    "kube-system",
+]
+.into_iter()
+.map(|ns| format!("metadata.namespace!={ns}"))
+.collect::<Vec<_>>()
+.join(",");
+let cfg = watcher::Config::default().fields(&ignoring_system_namespaces);
+```
+
+!!! note "Field Selector Limitations"
+
+    Due to [field-selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/) limitations, you cannot filter on arbitrary fields, nor can you do set complements (need to enumerate explicitly).
+
+If you find that field-selectors are too constrictive for your set of objects, the problem can generally be solved using by [explicitly labelling](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) the objects you care about and use label selectors instead:
+
+```rust
+let cfg = Config::default().labels("environment in (production, qa)");
+```
+
+### Watching Metadata Only
+
+Generally, **Kubernetes will return all fields** for an object when doing a watch, so while the main tool at our disposal to reduce networked traffic is to reduce the number of watched objects, there is one trick available:
+
+We can ask Kubernetes to only return [TypeMeta] (`.api_version` + `.kind`) + [ObjectMeta] (`.metadata`) in our watches and list calls using `Api::watch_metadata` or the stream analogue [metadata_watcher]; a drop-in replacement for [watcher]:
+
+```rust
+let cfg = watcher::Config::default().fields(&ignoring_system_namespaces);
+let stream = reflector(writer, metadata_watcher(api, cfg)).applied_objects();
+```
+
+While this will not work in all cases (if you need more data than metadata), it is a big improvement when it is feasible.
+
+Given metadata is generally only responsible from a **fraction of the data** in dense objects - and generally receives even 2x benefits for more sparse objects (see also [[#pruning-fields]]) - swapping to metadata_watchers can **significantly reduce reflector sizes** (memory footprint).
```

Comment:
```suggestion
Given that metadata is generally only responsible for a **fraction of the data** in dense objects - and generally receives even 2x benefits for more sparse objects (see also [[#pruning-fields]]) - swapping to metadata_watchers can **significantly reduce reflector sizes** (memory footprint).
```

Reply:
Probably also worth adding a note here about that this is primarily helpful when only the metadata is required (especially for watching owned objects and such?)..

If the first thing your reconciler does is `api.get()` then this would be less helpful.

Reply:
Yeah, I can see if there are some wording that can include that (merged the minor suggestion above). There are some legit cases for controllers from metadata watchers (not using owned/watched streams), but from my experience they are primarily label/annotation assigners.

If the reconciler calls `api.get()` - something people traditionally do for consistency - then you could probably still make the case that metadata watchers are helpful; you can run the watcher with `Any` semantic (because you are forcing an up-date anyway), and you also can do filtering of the watcher stream early so you only do full object fetches in the reconciler.